src/handleInteraction.ts

```diff
@@ -0,0 +1,192 @@
+ * typing indicators, ephemeral replies, partial fields,
+ * and error handling.
+ *
+ * The goal is for us devs to not have to worry about how exactly
+ * things get done when we're writing command handlers, only
+ * that what we say goes.
+ *
+ * @param logger The place to write system messages.
+ */
+export async function handleInteraction(
+	interaction: CommandInteraction,
+	logger: Console
+): Promise<void> {
+	// Don't respond to bots or ourselves
+	if (interaction.user.bot) return;
+	if (interaction.user.id === interaction.client.user.id) return;
+
+	logger.debug(`User ${logUser(interaction.user)} sent command: '${interaction.commandName}'`);
+
+	const command = allCommands.get(interaction.commandName);
+	if (!command) return;
+
+	logger.debug(
+		`Calling command handler '${command.name}' with options ${JSON.stringify(
+			interaction.options,
+			undefined,
+			2
+		)}`
+	);
+
+	let member: GuildMember | null;
+	if (interaction.inCachedGuild()) {
+		member = interaction.member;
+	} else {
+		member = (await interaction.guild?.members.fetch(interaction.user)) ?? null;
+	}
+
+	let channel: GuildTextBasedChannel | DMChannel | null;
+	if (interaction.channel?.type === ChannelType.DM && interaction.channel.partial) {
+		channel = await interaction.channel.fetch();
+	} else {
+		channel = interaction.channel;
+	}
+
+	const context: CommandContext = {
+		createdTimestamp: interaction.createdTimestamp,
+		user: interaction.user,
+		member,
+		guild: interaction.guild,
+		channel,
+		client: interaction.client,
+		interaction,
+		options: interaction.options.data,
+		logger,
+		async prepareForLongRunningTasks(ephemeral) {
+			try {
+				await interaction.deferReply({ ephemeral });
+			} catch (error) {
+				logger.error('Failed to defer reply to interaction:', error);
+			}
+		},
+		async replyPrivately(options, viaDM: boolean = false) {
+			if (viaDM) {
+				// We need to say *something* to the interaction itself, or Discord will think we died.
+				const content = ':paperclip: Check your DMs';
+				if (interaction.deferred) {
+					try {
+						await interaction.editReply(content);
+					} catch (error) {
+						logger.error('Failed to edit reply to interaction:', error);
+					}
+				} else {
+					try {
+						await interaction.reply({ content, ephemeral: true });
+					} catch (error) {
+						logger.error('Failed to reply to interaction:', error);
+					}
+				}
+			}
+			if (interaction.deferred && !viaDM) {
+				try {
+					if (typeof options === 'string') {
+						await interaction.followUp({ ephemeral: true, content: options });
+					} else {
+						await interaction.followUp({ ...options, ephemeral: true });
+					}
+				} catch (error) {
+					logger.error('Failed to follow up on interaction:', error);
+				}
+			} else {
+				const reply = await replyPrivately(interaction, options, viaDM);
+				if (reply === false) {
+					logger.info(`User ${logUser(interaction.user)} has DMs turned off.`);
```

Comment:
In this situation, maybe send an ephemeral reply to the command telling the user "I can't access your DMs! Please allow messages from server members and try again."

Reply:
Nevermind, you included that elsewhere.